File: typescript/src/resolver.ts
/** Resolve a DID to a verified identity. did:key offline; did:web over HTTPS
 *  with an injectable fetch. */

import * as signing from "./signing.ts";
import { InvalidIdentity, ResolutionError, UnsupportedMethod } from "./errors.ts";
import { type AgentIdentity, verifyIdentity } from "./identity.ts";

export const WELL_KNOWN = ".well-known/agent-identity.json";
const MAX_BYTES = 1_000_000;

export type Fetcher = (url: string) => Promise<unknown> | unknown;

const FETCH_TIMEOUT_MS = 15_000;
const MAX_REDIRECTS = 3;

export function didWebUrl(id: string): string {
  const parts = id.split(":");
  if (parts.length < 3 || parts[0] !== "did" || parts[1] !== "web") {
    throw new ResolutionError(`not a did:web identifier: ${id}`);
  }
  const domain = parts[2].replace("%3A", ":");
  const segments = parts.slice(3);
  // Reject path-traversal / empty segments so a DID cannot escape its own path.
  if (segments.some((s) => s === "" || s === "." || s === "..")) {
    throw new ResolutionError(`did:web has empty or traversal path segment: ${id}`);
  }
  return segments.length
    ? `https://${domain}/${segments.join("/")}/agent-identity.json`
    : `https://${domain}/${WELL_KNOWN}`;
}

/** Parse one IPv4 octet as libc inet_aton does: hex (0x..), octal (0..), decimal. */
function parseV4Octet(p: string): number | null {
  if (p === "") return null;
  const low = p.toLowerCase();
  if (low.startsWith("0x")) return /^0x[0-9a-f]+$/.test(low) ? parseInt(low, 16) : null;
  if (p.startsWith("0") && p.length > 1) return /^[0-7]+$/.test(p) ? parseInt(p, 8) : null;
  return /^[0-9]+$/.test(p) ? parseInt(p, 10) : null;
}

/** Parse an IPv4 literal in any encoding (dotted decimal/octal/hex, 1-to-4-part
 *  shorthand, bare integer) into a 32-bit value, matching getaddrinfo. */
function literalToV4(h: string): number | null {
  const parts = h.split(".");
  if (parts.length < 1 || parts.length > 4) return null;
  const nums = parts.map(parseV4Octet);
  if (nums.some((n) => n === null)) return null;
  const ns = nums as number[];
  const widths = { 1: [32], 2: [8, 24], 3: [8, 8, 16], 4: [8, 8, 8, 8] }[parts.length]!;
  const shifts = { 1: [0], 2: [24, 0], 3: [24, 16, 0], 4: [24, 16, 8, 0] }[parts.length]!;
  let v = 0;
  for (let i = 0; i < ns.length; i++) {
    if (ns[i] < 0 || ns[i] >= 2 ** widths[i]) return null;
    v += ns[i] * 2 ** shifts[i];
  }
  return v >>> 0;
}

function v4Blocked(v: number): boolean {
  const a = (v >>> 24) & 255;
  const b = (v >>> 16) & 255;
  if (a === 0 || a === 10 || a === 127) return true; // this-network, private, loopback
  if (a === 169 && b === 254) return true; // link-local incl. 169.254.169.254 metadata
  if (a === 172 && b >= 16 && b <= 31) return true; // private
  if (a === 192 && b === 168) return true; // private
  if (a === 100 && b >= 64 && b <= 127) return true; // CGNAT
  if (a >= 224) return true; // multicast + reserved + broadcast
  return false;
}

/** Expand an IPv6 literal (incl. :: compression and embedded IPv4) to 16 bytes. */
function ipv6ToBytes(h: string): Uint8Array | null {
  if (!h.includes(":")) return null;
  let str = h;
  // Fold an embedded IPv4 tail (e.g. ::ffff:127.0.0.1) into two hextets.
  const lastColon = str.lastIndexOf(":");
  const tail = str.slice(lastColon + 1);
  if (tail.includes(".")) {
    const v4 = literalToV4(tail);
    if (v4 === null) return null;
    str = str.slice(0, lastColon + 1) + ((v4 >>> 16) & 0xffff).toString(16) + ":" + (v4 & 0xffff).toString(16);
  }
  const halves = str.split("::");
  if (halves.length > 2) return null;
  let groups: string[];
  if (halves.length === 2) {
    const head = halves[0] ? halves[0].split(":") : [];
    const back = halves[1] ? halves[1].split(":") : [];
    const missing = 8 - head.length - back.length;
    if (missing < 1) return null;
    groups = [...head, ...Array(missing).fill("0"), ...back];
  } else {
    groups = str.split(":");
  }
  if (groups.length !== 8) return null;
  const bytes = new Uint8Array(16);
  for (let i = 0; i < 8; i++) {
    if (!/^[0-9a-fA-F]{1,4}$/.test(groups[i])) return null;
    const val = parseInt(groups[i], 16);
    bytes[i * 2] = (val >> 8) & 255;
    bytes[i * 2 + 1] = val & 255;
  }
  return bytes;
}

function v6Blocked(b: Uint8Array): boolean {
  // IPv4-mapped (::ffff:a.b.c.d) and IPv4-compatible: check the embedded v4.
  const firstTen = b.slice(0, 10).every((x) => x === 0);
  if (firstTen && b[10] === 0xff && b[11] === 0xff) {
    return v4Blocked(((b[12] << 24) | (b[13] << 16) | (b[14] << 8) | b[15]) >>> 0);
  }
  if (b.every((x) => x === 0)) return true; // ::
  if (firstTen && b[10] === 0 && b[11] === 0 && b[15] === 1) return true; // ::1 loopback
  if (b[0] === 0xfe && (b[1] & 0xc0) === 0x80) return true; // fe80::/10 link-local
  if ((b[0] & 0xfe) === 0xfc) return true; // fc00::/7 unique-local
  if (b[0] === 0xfe && (b[1] & 0xc0) === 0xc0) return true; // fec0::/10 site-local
  if (b[0] === 0xff) return true; // ff00::/8 multicast
  // 64:ff9b::/96 NAT64: block the prefix and the IPv4 it translates.
  if (b[0] === 0x00 && b[1] === 0x64 && b[2] === 0xff && b[3] === 0x9b) {
    return v4Blocked(((b[12] << 24) | (b[13] << 16) | (b[14] << 8) | b[15]) >>> 0);
  }
  return false;
}

/** Block SSRF targets across every IP encoding (loopback, private, link-local
 *  incl. the cloud metadata IP, internal-looking names). No DNS resolution, so a
 *  hostname pointing at an internal IP is still the caller's risk; use an
 *  allow-list for untrusted input. */
export function hostIsBlocked(host: string): boolean {
  const h = host.replace(/^\[|\]$/g, "").trim().toLowerCase();
  if (!h || h === "localhost" || /\.(local|internal|localhost)$/.test(h)) return true;
  const v6 = ipv6ToBytes(h);
  if (v6) return v6Blocked(v6);
  const v4 = literalToV4(h);
  if (v4 !== null) return v4Blocked(v4);
  return false; // a hostname; not resolved here
}

async function httpFetch(url: string, redirectsLeft = MAX_REDIRECTS): Promise<unknown> {
  if (!url.startsWith("https://")) throw new ResolutionError("did:web resolution requires HTTPS");
  const host = new URL(url).hostname;
  if (hostIsBlocked(host)) {
    throw new ResolutionError(`refusing to resolve a private/loopback host: ${host}`);
  }
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
  let res: Response;
  try {
    res = await fetch(url, {
      headers: { Accept: "application/json", "User-Agent": "identitykit/0.0.2" },
      redirect: "manual",
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timer);
  }
  // Re-check the SSRF guard on every redirect hop rather than following blindly.
  if (res.status >= 300 && res.status < 400) {
    const loc = res.headers.get("location");
    if (!loc) throw new ResolutionError("did:web redirect without a Location header");
    if (redirectsLeft <= 0) throw new ResolutionError("did:web exceeded redirect limit");
    return httpFetch(new URL(loc, url).toString(), redirectsLeft - 1);
  }
  if (!res.ok) throw new ResolutionError(`fetch failed: HTTP ${res.status}`);
  // Stream with a hard byte cap so a huge/slow body cannot exhaust memory.
  const reader = res.body?.getReader();
  if (!reader) throw new ResolutionError("did:web response has no body");
  const chunks: Uint8Array[] = [];
  let total = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    if (value) {
      total += value.length;
      if (total > MAX_BYTES) {
        await reader.cancel();
        throw new ResolutionError("identity document exceeds size limit");
      }
      chunks.push(value);
    }
  }
  const buf = Buffer.concat(chunks.map((c) => Buffer.from(c)));
  return JSON.parse(buf.toString("utf8"));
}

export interface DidKeyResolution {
  id: string;
  method: "key";
  public_key: string;
  note: string;
}

export async function resolve(id: string, opts: { fetch?: Fetcher } = {}): Promise<AgentIdentity | DidKeyResolution> {
  if (typeof id !== "string" || !id.startsWith("did:") || (id.match(/:/g) ?? []).length < 2) {
    throw new ResolutionError(`not a DID: ${id}`);
  }
  const method = id.split(":")[1];

  if (method === "key") {
    const pub = signing.publicFromDidKey(id);
    return {
      id,
      method: "key",
      public_key: signing.b64(pub),
      note: "did:key carries no hosted document; transport the signed AgentIdentity with it.",
    };
  }

  if (method === "web") {
    const url = didWebUrl(id);
    const doc = (await (opts.fetch ?? httpFetch)(url)) as AgentIdentity;
    if (!doc || typeof doc !== "object") throw new ResolutionError("fetched document is not a JSON object");
    if (doc.id !== id) throw new InvalidIdentity(`document id ${doc.id} does not match requested ${id}`, "id");
    if (!verifyIdentity(doc)) throw new InvalidIdentity("fetched document failed verification");
    return doc;
  }

  throw new UnsupportedMethod(`no resolver for did method ${method} (v0 supports key, web)`);
}

File: python/identitykit/resolver.py
"""Resolve an identifier to a verified identity.

did:key is offline and self-certifying: the key is in the id. did:web fetches a
signed AgentIdentity document over HTTPS from the domain's well-known location
and verifies it. The fetch is injectable so resolution can be tested offline and
so callers can supply their own transport, caching, or allow-list.
"""

from __future__ import annotations

import ipaddress
import json
import urllib.parse
import urllib.request
from typing import Callable, Dict, Optional

from . import signing
from .errors import InvalidIdentity, ResolutionError, UnsupportedMethod
from .identity import verify_identity

WELL_KNOWN = ".well-known/agent-identity.json"
_MAX_BYTES = 1_000_000  # refuse oversized documents

Fetcher = Callable[[str], Dict]


def did_web_url(id: str) -> str:
    """Map a did:web identifier to its HTTPS document URL (did:web spec path rules)."""
    parts = id.split(":")
    if len(parts) < 3 or parts[0] != "did" or parts[1] != "web":
        raise ResolutionError(f"not a did:web identifier: {id!r}")
    domain = parts[2].replace("%3A", ":")  # encoded port, if any
    segments = parts[3:]
    # Reject path-traversal / empty segments so a DID cannot escape its own path.
    if any(s in ("", ".", "..") for s in segments):
        raise ResolutionError(f"did:web has empty or traversal path segment: {id!r}")
    if segments:
        return f"https://{domain}/{'/'.join(segments)}/agent-identity.json"
    return f"https://{domain}/{WELL_KNOWN}"


def _parse_ipv4_octet(part: str) -> int:
    """Parse one octet as libc inet_aton does: hex (0x..), octal (0..), or decimal."""
    if part == "":
        raise ValueError("empty octet")
    low = part.lower()
    if low.startswith("0x"):
        return int(low, 16)
    if part.startswith("0") and len(part) > 1:
        return int(part, 8)
    return int(part, 10)


def _literal_to_ip(h: str):
    """Best-effort parse of an IP *literal* in any encoding the C resolver accepts
    (dotted decimal/octal/hex, 1-to-4-part shorthand, bare integer, IPv6,
    IPv4-mapped IPv6). Returns an ipaddress object, or None if `h` is a hostname.

    This closes the SSRF gap where ipaddress.ip_address() is strict but the actual
    socket connection's getaddrinfo() accepts forms like 2130706433 or 0x7f.0.0.1."""
    try:
        return ipaddress.ip_address(h)
    except ValueError:
        pass
    parts = h.split(".")
    if not (1 <= len(parts) <= 4):
        return None
    try:
        nums = [_parse_ipv4_octet(p) for p in parts]
    except ValueError:
        return None
    # inet_aton shorthand: the final part absorbs the remaining low-order bytes.
    widths = {1: [32], 2: [8, 24], 3: [8, 8, 16], 4: [8, 8, 8, 8]}[len(parts)]
    shifts = {1: [0], 2: [24, 0], 3: [24, 16, 0], 4: [24, 16, 8, 0]}[len(parts)]
    value = 0
    for num, width, shift in zip(nums, widths, shifts):
        if num < 0 or num >= (1 << width):
            return None
        value |= num << shift
    try:
        return ipaddress.ip_address(value)
    except ValueError:
        return None


def _ip_is_blocked(ip) -> bool:
    """Block anything not globally routable, after unwrapping IPv4-mapped IPv6.
    Covers loopback, private, link-local (incl. 169.254.169.254), reserved,
    multicast, unspecified, and site-local in one rule."""
    if ip.version == 6:
        if ip.ipv4_mapped is not None:
            ip = ip.ipv4_mapped
        # Deprecated site-local fec0::/10 is classed global by ipaddress; block it.
        elif (int(ip) >> 118) == 0x3FB:  # fec0::/10
            return True
    return not ip.is_global


def _host_is_blocked(host: str) -> bool:
    """Block obvious SSRF targets across every IP encoding, plus internal-looking
    names. Does not resolve DNS, so a hostname pointing at an internal IP is still
    the caller's risk — use an allow-list for untrusted input."""
    h = host.strip().strip("[]").lower()
    if not h or h == "localhost" or h.endswith((".local", ".internal", ".localhost")):
        return True
    ip = _literal_to_ip(h)
    if ip is not None:
        return _ip_is_blocked(ip)
    return False


class _GuardedRedirectHandler(urllib.request.HTTPRedirectHandler):
    """Re-apply the HTTPS + SSRF host checks on every redirect hop, so a public
    did:web document cannot 302 the resolver onto an internal host."""

    def redirect_request(self, req, fp, code, msg, headers, newurl):
        if not newurl.startswith("https://"):
            raise ResolutionError("did:web redirect to a non-HTTPS URL is blocked")
        host = urllib.parse.urlparse(newurl).hostname or ""
        if _host_is_blocked(host):
            raise ResolutionError(f"did:web redirect to a private/loopback host: {host!r}")
        return super().redirect_request(req, fp, code, msg, headers, newurl)


_opener = urllib.request.build_opener(_GuardedRedirectHandler())


def _http_fetch(url: str) -> Dict:
    if not url.startswith("https://"):
        raise ResolutionError("did:web resolution requires HTTPS")
    host = urllib.parse.urlparse(url).hostname or ""
    if _host_is_blocked(host):
        raise ResolutionError(f"refusing to resolve a private/loopback host: {host!r}")
    req = urllib.request.Request(url, headers={"Accept": "application/json", "User-Agent": "identitykit/0.0.2"})
    # https enforced above; redirects re-checked by _GuardedRedirectHandler.
    with _opener.open(req, timeout=15) as resp:  # nosec B310 # noqa: S310
        raw = resp.read(_MAX_BYTES + 1)
    if len(raw) > _MAX_BYTES:
        raise ResolutionError("identity document exceeds size limit")
    return json.loads(raw.decode("utf-8"))


def resolve(id: str, *, fetch: Optional[Fetcher] = None) -> Dict:
    """Resolve `id` to a verified result.

    did:web -> the fetched, signature-verified AgentIdentity document.
    did:key -> a minimal self-resolved stub carrying the controlling key. A
               did:key has no hosted document, so any richer AgentIdentity must
               travel alongside it and be checked with verify_identity().
    """
    if not isinstance(id, str) or id.count(":") < 2 or not id.startswith("did:"):
        raise ResolutionError(f"not a DID: {id!r}")
    method = id.split(":")[1]

    if method == "key":
        pub = signing.public_from_did_key(id)  # raises on a malformed key
        return {
            "id": id,
            "method": "key",
            "public_key": signing.b64(pub),
            "note": "did:key carries no hosted document; transport the signed AgentIdentity with it.",
        }

    if method == "web":
        url = did_web_url(id)
        doc = (fetch or _http_fetch)(url)
        if not isinstance(doc, dict):
            raise ResolutionError("fetched document is not a JSON object")
        if doc.get("id") != id:
            raise InvalidIdentity(f"document id {doc.get('id')!r} does not match requested {id!r}", field="id")
        if not verify_identity(doc):
            raise InvalidIdentity("fetched document failed verification")
        return doc

    raise UnsupportedMethod(f"no resolver for did method {method!r} (v0 supports key, web)")
